File: src/print_partner/core/scanner.py
"""Walk repo directories and collect STL parts."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from print_partner.config import settings
from print_partner.core.parsers import ParsedPart, parse_stl_path


@dataclass
class ScannedPart:
    relative_path: str
    filename: str
    match_key: str
    part_slug: str
    role: str
    quantity: int
    absolute_path: Path


def normalize_match_key(relative_path: str) -> str:
    return relative_path.replace("\\", "/").lower().strip("/")


def safe_repo_path(repo_root: Path, relative_path: str) -> Path | None:
    """Resolve path only if it stays under repo_root (reject ..)."""
    try:
        root = repo_root.resolve()
        candidate = (root / relative_path).resolve()
        candidate.relative_to(root)
        return candidate
    except (ValueError, OSError):
        return None


def is_under_data_repos(path: Path) -> bool:
    try:
        path.resolve().relative_to(settings.repos_dir.resolve())
        return True
    except ValueError:
        return False


def scan_repo(
    repo_root: Path,
    source_layer: str = "base",
    import_rules: list[str] | None = None,
) -> list[ScannedPart]:
    """
    Walk repo for STLs. import_rules: None = all (legacy), [] = none, else filter.
    """
    from print_partner.core.import_rules import path_matches_rules

    if not repo_root.is_dir():
        return []
    root = repo_root.resolve()
    if not is_under_data_repos(root) and settings.repos_dir.exists():
        try:
            root.relative_to(settings.repos_dir.resolve())
        except ValueError:
            pass

    parts: list[ScannedPart] = []
    for stl in root.rglob("*.stl"):
        if not stl.is_file():
            continue
        rel = str(stl.relative_to(root)).replace("\\", "/")
        if import_rules is not None and not path_matches_rules(rel, import_rules):
            continue
        parsed: ParsedPart = parse_stl_path(rel)
        parts.append(
            ScannedPart(
                relative_path=rel,
                filename=parsed.filename,
                match_key=normalize_match_key(rel),
                part_slug=parsed.part_slug,
                role=parsed.role.value,
                quantity=parsed.quantity,
                absolute_path=stl,
            )
        )
    return sorted(parts, key=lambda p: p.match_key)

File: web/packages/domain/src/scanner.ts
/** Walk repo directories and collect STL parts (ported from Python scanner.py). */

import { readdirSync, statSync } from "node:fs";
import { join, resolve } from "node:path";
import { pathMatchesRules } from "./import-rules.js";
import { parseStlPath } from "./parsers.js";
import type { NamingProfile } from "./stl-naming.js";

export type ScannedPart = {
  relativePath: string;
  filename: string;
  matchKey: string;
  partSlug: string;
  role: string;
  quantity: number;
  absolutePath: string;
};

export function normalizeMatchKey(relativePath: string): string {
  return relativePath.replace(/\\/g, "/").toLowerCase().replace(/^\/+|\/+$/g, "");
}

export function safeRepoPath(repoRoot: string, relativePath: string): string | null {
  try {
    const root = resolve(repoRoot);
    const candidate = resolve(root, relativePath);
    if (!candidate.startsWith(root + "/") && candidate !== root) return null;
    return candidate;
  } catch {
    return null;
  }
}

function walkStlFiles(dir: string, root: string, out: string[]): void {
  let entries: string[];
  try {
    entries = readdirSync(dir);
  } catch {
    return;
  }
  for (const name of entries) {
    const full = join(dir, name);
    let st: ReturnType<typeof statSync>;
    try {
      st = statSync(full);
    } catch {
      continue;
    }
    if (st.isDirectory()) {
      walkStlFiles(full, root, out);
    } else if (st.isFile() && name.toLowerCase().endsWith(".stl")) {
      out.push(full);
    }
  }
}

export function scanRepo(
  repoRoot: string,
  sourceLayer = "base",
  importRules: string[] | null = null,
  namingProfile?: NamingProfile | null,
): ScannedPart[] {
  void sourceLayer;
  let root: string;
  try {
    root = resolve(repoRoot);
    const st = statSync(root);
    if (!st.isDirectory()) return [];
  } catch {
    return [];
  }

  const stlPaths: string[] = [];
  walkStlFiles(root, root, stlPaths);

  const parts: ScannedPart[] = [];
  for (const stl of stlPaths) {
    const rel = stl.slice(root.length + 1).replace(/\\/g, "/");
    if (importRules != null && !pathMatchesRules(rel, importRules)) continue;
    const parsed = parseStlPath(rel, namingProfile);
    parts.push({
      relativePath: rel,
      filename: parsed.filename,
      matchKey: normalizeMatchKey(rel),
      partSlug: parsed.partSlug,
      role: parsed.role,
      quantity: parsed.quantity,
      absolutePath: stl,
    });
  }

  return parts.sort((a, b) => a.matchKey.localeCompare(b.matchKey));
}

export function listStlRelativePaths(repoRoot: string): string[] {
  let root: string;
  try {
    root = resolve(repoRoot);
    if (!statSync(root).isDirectory()) return [];
  } catch {
    return [];
  }
  const stlPaths: string[] = [];
  walkStlFiles(root, root, stlPaths);
  return stlPaths
    .map((p) => p.slice(root.length + 1).replace(/\\/g, "/"))
    .sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
}
